### PR TITLE
Add OwnerReverseLookup modaility

### LINKED ISSUE #93: modality to opt in to owner reporting
Add a modality to opt in to owner reporting (reverse lookup), defaulted to false - existing clients will be migrated to "on" .  What happens when a client with this turned off receives a receipt from a client with this turned on? 

May want to put a hard cap of NFTs that can be minted?  this allows us to limit the size of the data we have to store.  Higher costs are driven by the number of NFTs we want to allow.  Blockchains are not meant to store large data.  

Moving to a more naive implementation - allocate max space up front (large cost), lower cost on per-mint basis.

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

- Added a new `ReportingMode` modality that dictates whether the contract will track ownership via pages or not
- Added a new `register_owner` entrypoint allowing **contract users** to pay the cost of page table allocation upfront, rather than getting bit by a one time fee on first minted / transferred per owner.
- Added a hard limit of `1_000_000` tokens (1 million) per CEP-78 instance (_technically_ a breaking change, but analysis of mainnet found no existing contract even remotely close to that many tokens)
- Increased page size from `10` to `1000`
- Updated the mint session WASM to invoke the register the sender of the WASM before the mint entrypoint is invoked
- Expanded test suite to account for the new modality

Refer to the README which describes the newly added modality and entry point.

Closes #93

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -239,6 +239,23 @@ provides two options:
...
+2. `Report`: The reporting and receipt functionality is supported.
+
+| ReportingMode | u8  |
+|---------------|-----|
+| Report        | 0   |
+| NoReport      | 1   |
+
+This modality is an optional installation parameter and will default to the `NoReport` mode if not provided. However, this
+mode cannot be changed once the contract has been installed. The mode is set by passing a `u8` value to the `reporting_mode` runtime argument.
+
```
Comment (EdHastingsCasperAssociation) [inline README.md]: let's rename this to `OwnerReverseLookupMode`, and rename the variants to `Complete` and `None`...leaving the door open later to a non-binary choice like `TransferredOnly` or sth like that

also, important, add clarification around why having receipts issued and the ability to do a reverse look up for all owned by owner is very useful, and give some suggestion as to which use cases it does and does not pertain to. 

then also clarify the down side, which is that turning this option on results in significantly higher gas costs for all uses of mint and transfer

i.e. communicating the WHY is much more important than the WHAT

Comment (darthsiroftardis) [reply README.md]: @EdHastingsCasperLabs I am going to go with `NoLookup` as opposed to `None`, just to keep it distinct from the `Option::None`

File: contract/src/main.rs [hunk 2]
```diff
@@ -583,23 +543,79 @@ pub extern "C" fn mint() {
...
+
+        let mut page = if !page_table[page_table_entry as usize] {
+            let _ = core::mem::replace(&mut page_table[page_table_entry as usize], true);
+            storage::dictionary_put(page_table_uref, &owned_tokens_item_key, page_table);
+            // match storage::dictionary_get::<Vec<bool>>(page_uref,
+            // &page_item_key).unwrap_or_revert() {         Some(single_page) =>
+            // single_page,         None => {
+            //             vec![false; PAGE_SIZE as usize]
+            //         }
+            //     }
```
Comment (EdHastingsCasperAssociation) [inline contract/src/main.rs]: delete dead code

File: contract/src/main.rs [hunk 3]
```diff
@@ -1364,6 +1379,49 @@ pub extern "C" fn updated_receipts() {
     runtime::ret(CLValue::from_t(updated_receipts).unwrap_or_revert())
 }
 
+#[no_mangle]
+pub extern "C" fn register_owner() {
+    if let ReportingMode::NoReport = utils::get_reporting_mode() {
+        runtime::revert(NFTCoreError::InvalidReportingMode)
+    }
+
```
Comment (EdHastingsCasperAssociation) [inline contract/src/main.rs]: add code comments for and to this entry point explaining its use and helping people understand what the code is doing

**Review by EdHastingsCasperAssociation — APPROVED** [review]
please address nits

File: contract/src/main.rs (hunk 3 above)
Comment (darthsiroftardis) [reply contract/src/main.rs]: Aye sir

**Comment by mssteuer** [comment]
IMHO the hard limit of 1M tokens should be considered in the context of:
- you have at least one enterprise client whose intention it is to mint 10s of millions of NFTs under the same contract, in my understanding
- for a mass market standard, this really may be a problem. We have hopefully mass market games coming online that intend to use this standard. Does that mean there can only be 1M items total in these game? If these games are as successful as we expect and hope them to be, they'll blow through that in the first weeks

**Comment by darthsiroftardis** [comment]
> * or a mass market standard, this really may be a problem. We have hopefully mass market games coming online that intend to use this standard. Does that mean there can only be 1M items total in these game? If these games are as successful as we expect and hope them to be, they'll blow through that in the first weeks

Yes, however, we expect that in those cases the installing account with install multiple CEP-78 instances, and we've also changed the installation process such that the named key entries won't be overridden when doing so. We're also in the process of updating the documentation to reflect that. However given the amount of changes needed to be documented that will be a separate PR

File: README.md (hunk 1 above)
Comment (darthsiroftardis) [reply README.md]: > 

@ACStoneCL Please account for this readme PR, I think you and I have discussed the WHY, just pointing it out here AGAIN

File: README.md [hunk 4]
```diff
@@ -274,6 +291,7 @@ The following are the optional parameters that can be passed in at the time of i
...
 * `"contract_whitelist"`: The contract whitelist is a list of contract hashes that specifies which contracts can call the `mint()` entrypoint to mint NFTs. This is an optional parameter which will default to an empty whitelist. This value can be changed via the `set_variables` post installation. If the whitelist mode is set to locked, a non-empty whitelist must be passed; else, installation of the contract will fail.
 * `"burn_mode"`: The [`BurnMode`](#burnmode) modality dictates whether minted NFTs can be burnt. This is an optional parameter and will allow tokens to be burnt by default. This parameter cannot be changed once the contract has been installed.
+* `"owner_reverse_lookup_mode"`: The [`OwnerReverseLookupMode`](#reportingmode) modality dictates whether the lookup for owners to token identifiers is available. This is an optional parameter will not provide the lookup by default. This parameter cannot be changed once the contract has been installed.
```
Comment (ACStone-MTS) [inline README.md]:
```suggestion
* `"owner_reverse_lookup_mode"`: The [`OwnerReverseLookupMode`](#reportingmode) modality dictates whether the lookup for owners to token identifiers is available. This is an optional parameter and will not provide the lookup by default. This parameter cannot be changed once the contract has been installed.
```

**Pull request merged**